### ISSUE TITLE
Swift compatibility

### ISSUE THREAD

**Issue by nanoxd** [bryankeller/BLKFlexibleHeightBar#1]
Hi,

I was trying to get this setup on a Swift project and noticed that it wouldn't allow me to use the tableView delegate using `BLKDelegateSplitter` or through `myBar.behaviorDefiner)

Inspecting the object leads me to believe that it doesn't understand how to cast it as UITableView delegate on either end. Here are the forms I've tried:

* `tableView.delegate = self` displays the bar but does not allow the bar to snap
* `tableView.delegate = behavior as UITableViewDelegate` crashes immediately and yields warning of "Treating a forced downcast to UITableViewDelegate as optional will never produce nil"
* `tableView.delegate = behavior as? UITableViewDelegate` loads up the view but the delegate is `nil`
* `tableView.delegate = (behavior as UITableViewDelegate)` silences the second warning but crashes because of nil value
* `tableView.delegate =  BLKDelegateSplitter(firstDelegate: behavior, secondDelegate: self) as? UITableViewDelegate` also yields nil

I tried recreating the Objective-C demo for the Square style and got the same results. Do you intend to add Swift support in the near future?

**Comment by bryankeller** [comment]
Hi nonoxd,
I'll look into this issue this weekend. I definitely want this to be useable in swift projects, so expect compatibility in the near future.

**Comment by nanoxd** [comment]
Thanks, let me know if you need anything from me.

**Comment by rabzu** [comment]
Hi,
I'm trying to use your project in swift. I downloaded the project using cocoapods and then set up a bridging header with:

#import "BLKFlexibleHeightBar.h"

This compiles fine. Editor in my ViewController sees BLKFlexibleHeightBar class. However, when I start to create BLKFlexibleHeightBar like below:

        var myBar = BLKFlexibleHeightBar(frame: frame)

I get an error:
    Use of unresolved identifier 'BLKFlexibleHeightBar'

Can anyone help because I tried everything.

Thanks,

**Comment by nanoxd** [comment]
@rabzu You need to specify `import BLKFlexibleHeightBar` at the top of the swift file you want to import it into.

**Comment by rabzu** [comment]
@nanoxd thank you for your response. When I try that I get error:
No Such Module 'BLKFlexibleHeightBar'

:(

**Comment by nanoxd** [comment]
@rabzu Make sure you use `#import <BLKFlexibleHeightBar/BLKFlexibleHeightBar.h>` in the bridging header if you are using cocoapods

**Comment by rabzu** [comment]
I just tried that but the No Such Module 'BLKFlexibleHeightBar'  error remains. :(

**Comment by nanoxd** [comment]
@rabzu Have you tried cleaning the build folder? Option+Shift+Command+K

**Comment by rabzu** [comment]
Yes, I even deleted Derived Data manually. But still not luck. Am I not doing do 
![screen shot 2015-03-13 at 16 54 48](https://cloud.githubusercontent.com/assets/6705064/6643131/d2cada7c-c9a1-11e4-8558-20f7a5ba30ad.png)

**Comment by nanoxd** [comment]
What does your podfile look like?

**Comment by rabzu** [comment]
This is my Podfile and Podfile.lock
![screen shot 2015-03-13 at 17 00 30](https://cloud.githubusercontent.com/assets/6705064/6643224/84f2fb76-c9a2-11e4-913a-091275647bb3.png)

**Comment by nanoxd** [comment]
@rabzu There's an update for cocoapods 0.36 stable. To use cocoapods with Swift, you must add `use_frameworks!` to your Podfile

**Comment by rabzu** [comment]
@nanoxd thank you sow much!!!! I'm new to this

**Comment by nanoxd** [comment]
No problem!
![high-five-yourself](https://cloud.githubusercontent.com/assets/1889575/6643535/36974014-c96a-11e4-8cd5-8e06f52c350c.gif)

**Comment by bryankeller** [comment]
I think the proper solution is to subclass the behavior definer you want to use and make it conform to UITableViewDelegate (or whatever other delegate protocol is appropriate in your scenario). Here's the basic idea if you're using CocoaPods:

Add a new Objective-C Cocoa Touch class and call it something appropriate (and probably far too wordy and descriptive) like `SquareCashStyleBehaviorDefinerForTableView`. The target membership should be
![screen shot 2015-03-14 at 1 49 55 am](https://cloud.githubusercontent.com/assets/746571/6650509/6d64fd20-c9ec-11e4-9ebf-e500ccc38ded.png)
(Note the "public")

Next, in `Pods-BLKFlexibleHeightBar-umbrella.h`, add

```objc
#import "SquareCashStyleBehaviorDefinerForTableView.h"
```

Finally, back in your Swift project, use this code to setup your behavior definer:
```swift
myBar.behaviorDefiner = SquareCashStyleBehaviorDefinerForTableView()
self.tableView?.delegate = myBar.behaviorDefiner as? UITableViewDelegate
```

I'll leave this open for a while just to see if anyone who has a bit more experience with Swift would like to chime in. For now, let me know if that works for you.

**Comment by rabzu** [comment]
I tried this and it works. Thanks

**Comment by bryankeller** [comment]
Happy to hear! 👍

**Comment by rabzu** [comment]
Similar approach also works with BLKDelegateSplitter.

**Comment by nanoxd** [comment]
@bryankeller Thanks for the update. I didn't get a chance to try it out yet but I'll report back with my findings when I do.

**Comment by iRambax** [comment]
There is also another simple way to fix tableView delegate issue. Just changed it from:

@interface BLKDelegateSplitter : NSObject <NSObject>
to:
@interface BLKDelegateSplitter : NSObject <UITableViewDelegate>

**Comment by rabzu** [comment]
@iRambax that does not work for me

**Comment by iRambax** [comment]
@rabzu any error message? Here is the way I use it on my Swift project which works fine:

let behaviorDefiner = SquareCashStyleBehaviorDefiner()
...
delegateSplitter = BLKDelegateSplitter(firstDelegate: behaviorDefiner, secondDelegate: self)
tableView.delegate = delegateSplitter

**Comment by rabzu** [comment]
I made work now, thanks :+1:

**Comment by bryankeller** [comment]
Glad you guys got it working!

**Comment by rtrivedi** [comment]
I'm trying to get this to work for my collectionview in swift but the header doesn't seem to shrink when I scroll. Here is a snippet of my code, any help would be appreciated!:

![screen shot 2015-03-23 at 10 28 50 pm](https://cloud.githubusercontent.com/assets/4014683/6796215/2b91b326-d1ac-11e4-866f-5f06fadd4ec1.png)

**Comment by rtrivedi** [comment]
@rabzu Do you have a working swift project with the SquareCashStyle header?

**Comment by rtrivedi** [comment]
@iRambax looks like you also mention that you've got this to work in Swift. Any idea how I can make this work?

Thanks!

Ronak

**Comment by iRambax** [comment]
@rtrivedi in your Swift project just make sure self.collectionView.delegate is not nil by following my two previous posts. You do not need to change anything else.

**Comment by rtrivedi** [comment]
Hi all, I just wanted to follow up so that other people in a similar situation can benefit from my investigations. It turns out, what I needed to do to get this to work is wrap my tableview in a UIViewController. That was the key. Otherwise adding the header to the self.view (when using a tableviewcontroller) didn't work.

Finally, I made the following additions to my uiview controller:

    func scrollViewDidEndDecelerating(scrollView: UIScrollView) { // called when scroll view grinds to a halt
    header.behaviorDefiner.scrollViewDidEndDecelerating(scrollView)
    }
    // called on finger up if the user dragged. decelerate is true if it will continue moving afterwards
    func scrollViewDidEndDragging(scrollView: UIScrollView, willDecelerate decelerate: Bool) {
    header.behaviorDefiner.scrollViewDidEndDragging(scrollView, willDecelerate: decelerate)
    }
    
    func scrollViewDidScroll(scrollView: UIScrollView) {
    let offset = scrollView.contentOffset
    let size = scrollView.contentSize
    
    if (size.height > 0)    // don't div by zero
    {
    let progress = 100 - (offset.y * 100 / size.height)
    
    header.progress = progress
    }
    header.behaviorDefiner.scrollViewDidScroll(scrollView)
    }

**Comment by rabzu** [comment]
My tableview delegate does not work when using BLKDelegateSplitter

@iRambax  when you define a delegate splitter in swift like this

let behaviorDefiner = SquareCashStyleBehaviorDefiner()
delegateSplitter = BLKDelegateSplitter(firstDelegate: behaviorDefiner,  secondDelegate: self)
tableView.delegate = delegateSplitter

does the tableview delegate work?

**Comment by rabzu** [comment]
For swift people, I think this is the best option to make it compatible with swift:
extension BLKDelegateSplitter: UITableViewDelegate{}

**Comment by Urkman** [comment]
It would be great to have a Swift Sample Project :)

**Comment by rabzu** [comment]
I can do that, but where should I upload?

**Comment by bryankeller** [comment]
You can submit a pull request. Make sure the example is simple and easy to understand. Model it off of the example projects included or even the example in the README.md.

This might also be a good place to use a delegate splitter so that people can see how that works.

**Comment by szoszig** [comment]
I checked out your swift sample, and I also implemented it into my app.
Using the extension solves the signed error, but the delegate doesn't work.
I cannot set neither sectionHeaderHeight, nor rowHeight, because it crashes.

"Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[CALayer tableView:heightForHeaderInSection:]: unrecognized selector sent to..."

Any advice?
Thx
